// string-validate/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLongText = exports.isShortText = exports.isPassword = exports.isOneOf = exports.isEqual = exports.isNumber = exports.isInt = exports.isPhone = exports.isEmail = exports.isRegex = exports.isLength = exports.isRequired = undefined;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var REGEX_EMAIL = /^((([a-z]|\d|[!#$%&'*+\-/=?^_`{|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+(\.([a-z]|\d|[!#$%&'*+\-/=?^_`{|}~]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])+)*)|((\x22)((((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(([\x01-\x08\x0b\x0c\x0e-\x1f\x7f]|\x21|[\x23-\x5b]|[\x5d-\x7e]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(\\([\x01-\x09\x0b\x0c\x0d-\x7f]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))))*(((\x20|\x09)*(\x0d\x0a))?(\x20|\x09)+)?(\x22)))@((([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|\d|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))\.)+(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])|(([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])([a-z]|\d|-|\.|_|~|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])*([a-z]|[\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])))$/i; /* eslint no-control-regex: "off" */

var REGEX_INT = /^\d+$/;
var REGEX_FLOAT = /^\d+(\.\d+)?$/;
var REGEX_PHONE = /^(?=.{2,50}$)[/\-.()+\s\d–]*\d{2,}[/\-.()+\s\d–]*$/;

var isRequired = exports.isRequired = function isRequired(value) {
  if (typeof value === 'undefined' || value === null) return false;
  if (Array.isArray(value)) return value.length !== 0;

  switch (typeof value === 'undefined' ? 'undefined' : (0, _typeof3.default)(value)) {
    case 'string':
      return (0, _utils.unicodeLength)(value) !== 0;
    case 'object':
      return JSON.stringify(value) !== '{}';
    case 'boolean':
      return value;
    case 'number':
      return true;
    default:
      return Boolean(value);
  }
};

var isLength = exports.isLength = function isLength() {
  var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var min = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  var max = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : Infinity;

  if (typeof value === 'undefined' || value === null) return false;
  var length = Array.isArray(value) ? value.length : (0, _utils.unicodeLength)(value);
  return max >= length && length >= min;
};

var isRegex = exports.isRegex = function isRegex(value, regex) {
  return regex.test(value);
};
var isEmail = exports.isEmail = function isEmail(value) {
  return isRegex(value, REGEX_EMAIL);
};
var isPhone = exports.isPhone = function isPhone(value) {
  return isRegex(value, REGEX_PHONE);
};
var isInt = exports.isInt = function isInt(value) {
  return isRegex(value, REGEX_INT);
};
var isNumber = exports.isNumber = function isNumber(value) {
  return isRegex(value, REGEX_FLOAT);
};
var isEqual = exports.isEqual = function isEqual(value, prop) {
  return value === prop;
};
var isOneOf = exports.isOneOf = function isOneOf(value) {
  for (var _len = arguments.length, props = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    props[_key - 1] = arguments[_key];
  }

  return props.includes(value);
};

// Common enough to be share-worthy
var isPassword = exports.isPassword = function isPassword(value) {
  return isLength(value, 8);
};
var isShortText = exports.isShortText = function isShortText(value) {
  return isLength(value, 2, 250);
};
var isLongText = exports.isLongText = function isLongText(value) {
  return isLength(value, 2, 5000);
};
